pkg/middleware/auth.go

```diff
@@ -146,6 +148,49 @@ func Auth(options *AuthOptions) web.Handler {
 	}
 }
 
+func RoleAppPluginAuthAndSignedIn(ps pluginstore.Store) web.Handler {
+	return func(c *contextmodel.ReqContext) {
+		if !c.IsSignedIn {
+			notAuthorized(c)
+			return
+		}
+
+		pluginID := web.Params(c.Req)[":id"]
+		p, exists := ps.Plugin(c.Req.Context(), pluginID)
+		if !exists {
+			c.JsonApiErr(http.StatusNotFound, "Plugin not found", nil)
+			return
+		}
+
+		found := false
+		allowed := false
+		path := filepath.Clean(c.Req.RequestURI)
+		for _, i := range p.Includes {
+			if i.Type != "page" {
+				continue
+			}
+
+			if filepath.Clean(i.Path) == path {
+				found = true
+				if i.Role == "" || c.HasRole(i.Role) {
+					allowed = true
+				}
+				break
+			}
+		}
+
+		if !found {
+			// This isn't an API request, so we probably shouldn't return a JSON error.
```

Comment:
TODO: Confirm

Reply:
See  https://github.com/grafana/grafana/blob/2c0c51a0df31bbc27ada3bcf3b2ea5584ef17846/pkg/api/index.go#L225-L238

Guess there's no way to say "skip route handler and let it end up at `NotFoundHandler`"?

Reply:
Yeah I just didn't want to bring the `HTTPServer` into the mix 😬